### PR TITLE
fix(broker): throw recoverable exception when LogStorageSteps fails due to newer transitions

### LINKED ISSUE #9586: llegalStateException: Not leader of partition 2
Related to #9040

We see errors like: [IllegalStateException: Not leader of partition 2](https://console.cloud.google.com/errors/detail/CPSxlN7o1JWZDg;service=zeebe;version=medic-cw-16-9910a4f121-benchmark;time=P7D?project=zeebe-io)

Which seems to be more or less expected if I take a look at the log.

[Log:](https://console.cloud.google.com/logs/query;query=%0A%0AlogName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.location%3D%22europe-west1-b%22%0Aresource.labels.cluster_name%3D%22zeebe-cluster%22%0Aresource.labels.namespace_name%3D%22medic-cw-16-9910a4f121-benchmark%22%0Aresource.labels.project_id%3D%22zeebe-io%22%0Aresource.labels.pod_name%3D%22medic-cw-16-9910a4f121-benchmark-zeebe-2%22%0Aresource.labels.container_name%3D%22zeebe%22%0AjsonPayload.context.partitionId%3D%222%22;timeRange=2022-05-03T22:48:42.982Z%2F2022-05-03T23:48:42.982Z;pinnedLogId=2022-05-03T23:18:12.982948166Z%2F9780n9n6ltcaoeg0;cursorTimestamp=2022-05-03T23:18:12.999252382Z?project=zeebe-io
)

```
2022-05-04 01:18:12.979 CEST
zeebe
Transition to LEADER on term 36 starting
2022-05-04 01:18:12.979 CEST
zeebe
Transition to LEADER on term 36 - transitioning LogStorage
2022-05-04 01:18:12.979 CEST
zeebe
RaftServer{raft-partition-partition-2}{role=LEADER} - Accepted VoteRequest{term=37, candidate=1, lastLogIndex=970255264, lastLogTerm=36}: candidate's log is up-to-date
2022-05-04 01:18:12.981 CEST
zeebe
Transition to FOLLOWER on term 37 requested.
2022-05-04 01:18:12.982 CEST
zeebe
Partition role transitioning from LEADER to FOLLOWER in term 37
2022-05-04 01:18:12.982 CEST
zeebe
Received cancel signal for transition to LEADER on term 36
2022-05-04 01:18:12.982 CEST
zeebe
Failed to install partition 2

2022-05-04 01:18:12.983 CEST
zeebe
RaftServer{raft-partition-partition-2} - Found leader 1
2022-05-04 01:18:12.994 CEST
zeebe
ZeebePartition-2 failed, marking it as unhealthy: Partition-2{status=UNHEALTHY, issue=ZeebePartition-2{status=UNHEALTHY, issue='Initial state'}}
2022-05-04 01:18:12.994 CEST
zeebe
Detected 'UNHEALTHY' components. The current health status of components: [ZeebePartition-2{status=UNHEALTHY, issue='Services not installed'}, raft-partition-partition-2{status=HEALTHY}]
```

Logs are stored as well under https://drive.google.com/drive/u/0/folders/1vYJilfkLRlrF9CqTAUwRT9dLeDqTZULb 

Exception:

```
java.lang.IllegalStateException: Not leader of partition 2
	at io.camunda.zeebe.broker.system.partitions.impl.steps.LogStoragePartitionTransitionStep.lambda$createWritableLogStorage$1(LogStoragePartitionTransitionStep.java:111) ~[zeebe-broker-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at java.util.Optional.orElseGet(Optional.java:364) ~[?:?]
	at io.camunda.zeebe.broker.system.partitions.impl.steps.LogStoragePartitionTransitionStep.createWritableLogStorage(LogStoragePartitionTransitionStep.java:107) ~[zeebe-broker-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.broker.system.partitions.impl.steps.LogStoragePartitionTransitionStep.buildAtomixLogStorage(LogStoragePartitionTransitionStep.java:83) ~[zeebe-broker-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.broker.system.partitions.impl.steps.LogStoragePartitionTransitionStep.transitionTo(LogStoragePartitionTransitionStep.java:50) ~[zeebe-broker-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.broker.system.partitions.impl.PartitionTransitionProcess.lambda$proceedWithTransition$1(PartitionTransitionProcess.java:80) ~[zeebe-broker-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.util.sched.ActorJob.invoke(ActorJob.java:79) ~[zeebe-util-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.util.sched.ActorJob.execute(ActorJob.java:44) ~[zeebe-util-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.util.sched.ActorTask.execute(ActorTask.java:122) ~[zeebe-util-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.util.sched.ActorThread.executeCurrentTask(ActorThread.java:97) ~[zeebe-util-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.util.sched.ActorThread.doWork(ActorThread.java:80) ~[zeebe-util-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
	at io.camunda.zeebe.util.sched.ActorThread.run(ActorThread.java:189) ~[zeebe-util-8.1.0-SNAPSHOT.jar:8.1.0-SNAPSHOT]
```

_Originally posted by @Zelldon in https://github.com/camunda/zeebe/issues/9040#issuecomment-1122493092_

### PR DESCRIPTION
**Pull request by deepthidevaki**
## Description

Since the role transitions in raft and zeebe partitions are asychronous, it is possible that while zeebe partition is installing leader services the raft has already transitioned to follower role. In this case, LogStorageStep fails because it cannot get the appender as the raft is not in the leader role anymore. Previously this fails the installation and the partition goes to an unhealthy state. But since this is an expected case, we could ignore it and continue with the installation of the new role. To enable this, we now throw a recoverable exception which won't be considered as a fatal error by ZeebePartition.

## Related issues

closes #9586 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Review by ChrisKujawa — APPROVED** [review]
Thanks @deepthidevaki :+1: sorry for the delay

File: broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/LogStoragePartitionTransitionStep.java [hunk 1]
```diff
@@ -120,18 +120,18 @@ private Either<Exception, AtomixLogStorage> checkAndCreateAtomixLogStorage(
 
     if (raftTerm != targetTerm) {
       return left(
-          new LogStorageTermMissmatchException(targetTerm, raftTerm, context.getPartitionId()));
+          new NoLeaderException(
```
[the remark this answers is not in the document]
Comment (deepthidevaki) [reply broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/steps/LogStoragePartitionTransitionStep.java:123]: I can change it to "NotLeaderException".

I think that `NotLeader` is more informative than `LogStorageTermMismatch`. And both cases (the one fixed in this PR and this term mismatch) the root cause is same - which is there was a concurrent role change in raft.

**Comment by deepthidevaki** [comment]
bors merge

**Comment by deepthidevaki** [comment]
bors merge

**Pull request merged**